frontend/src/metabase/parameters/utils/dashboards.ts

```diff
@@ -136,7 +138,11 @@ function buildFieldFilterUiParameter(
   const mappingsForParameter = mappings.filter(
     mapping => mapping.parameter_id === parameter.id,
   );
-  const mappedFields = mappingsForParameter.map(mapping => {
+  const uniqueMappingsForParameters = _.uniq(mappingsForParameter, mapping =>
+    JSON.stringify(mapping.target),
```

Comment:
An array-bound function using serialization per item immediately raises a red flag. It might become painfully slow at some point, probably even now. 

I don't fully understand the mapping semantics but would picking a bunch of ids and hashing them be enough to determine uniqueness? Calculating unique keys, hashing them and comparing numbers must be faster than serializing entire objects and comparing resulting strings.

Let's say we take a parameter:
```
{
    "parameter_id": "-1558815535",
    "target": [
        "dimension",
        [
            "field",
            11,
            {
                "source-field": 21
            }
        ]
    ],
    "card_id": "G__140569",
    "dashcard_id": "G__140589",
    "card": {}
}
```

Then we pick the fields into an array in a deterministic order, `[-1558815535, "G__140569", "G__140589"]`, then take an `md5` or any other hash of it or a string made of it.

WDYT?

Reply:
good point.

From what I understood, `mappingsForParameter` array is used for generating some objects based on the `target` field, that's why we had a lot of duplicated results, but further calculations are 100x slower that's why we want to reduce the size of this array.

I don't understand the logic fully as well, so I might be wrong here, but that idea was suggested by @kulyk and I believe him 😄

Reply:
We found that my suggestion won't fix anything as the most intensive calculations are happening elsewhere.

Reply:
A bit of context on why we're doing this: dashcards have `parameter_mappings` which associate a dashcard with many dashboard filters. For example, let's say we have a dashboard that looks roughly like that:

```js
{
  name: "Dashboard",
  parameters: [{
     id: "param-1",
     name: "Date",
     type: "date/all-options"
  }],
  dashcards: [{
     id: 1,
     card_id: 1,
     parameter_mappings: [{
       // if we've added other series to this dashcard (`series` prop),
       // it's important to set `card_id`
       card_id: 1,

       parameter_id: "param-1",

       // Let's pretend it's a reference to a "Created At" field
       target: [ "dimension", [ "field", 10, null ] ]
     }]
  }]
}
``` 

A parameter mapping object here defines how the daschard data is connected to the dashboard filter. As a result, once we change the filter value, we'll add a filter by the `[ "field", 10, null ]` field (specified in `parameter_mappings[0].target`)

Now imagine a dashboard with several filters and a few dozen of dashcards. They often would have queries to same tables, and will have very similar parameter mappings (imagine we have a datetime filter, there'll be a large chunk of dashcards connecting the same "CREATED_AT" column or something). In that case, `buildFieldFilterUiParameter` will loop over dozens of identical parameter mappings, instantiate a bunch of nearly identical MLV2 queries, etc. It doesn't make any sense, so we can just de-dupe this list upfront 

Instantiaing a MLv1 question is cheap, so we didn't bother about optimizing that step. But once we started constructing MLv2 queries inside this loop, the slowness started to stack up

Reply:
> We need to de-duplicate `mappingsForParameter`, so we only have objects with unique `target` values. IMO, we should ideally have a MLv2 function comparing these targets. There's also [`compareMappingOptionTargets` function](https://github.com/metabase/metabase/blob/4ac58f9ded42c405bc43e9ca098d98116fa43337/frontend/src/metabase-lib/parameters/utils/targets.ts#L71), but I'm not sure if it's correct (it's comparing fields by IDs).

I used MLv2 version to compare option targets and with currently available MLv2 methods it's even slower

Reply:
😢

Reply:
we have an area for improvement in MLv2 way, e.g. we can cache `visibleColumns`, but it doesn't looks like MLv2 way of working with data for me

Reply:
@kulyk what do you think if we keep this `JSON.stringify` thing with a comment? I'm working on a PR to improve calculations